internal/runners/branch/tree.go

```diff
@@ -0,0 +1,192 @@
+		if branch.Tracks == nil {
+			continue
+		}
+
+		// Check that this branch is a child branch and recursively build its tree
+		if _, ok := t[branchNode{branch.Label, branch.BranchID.String()}]; ok {
+			t[branchNode{branch.Label, branch.BranchID.String()}] = buildBranchTree(branch, branches)
+		}
+	}
+	return t
+}
+
+func getRootBranches(branches mono_models.Branches) mono_models.Branches {
+	var rootBranches mono_models.Branches
+	for _, branch := range branches {
+		// Account for forked projects where the root branches contain
+		// a tracking ID that is not in the current project's branches
+		if branch.Tracks != nil && containsBranch(branch.Tracks, branches) {
+			continue
+		}
+		rootBranches = append(rootBranches, branch)
+	}
+	return rootBranches
+}
+
+func containsBranch(id *strfmt.UUID, branches mono_models.Branches) bool {
+	for _, branch := range branches {
+		if branch.BranchID.String() == id.String() {
+			return true
+		}
+	}
+	return false
+}
+
+func getChildren(branch *mono_models.Branch, branches mono_models.Branches) tree {
+	children := make(tree)
+	if branch == nil {
+		return children
+	}
+
+	for _, b := range branches {
+		if b.Tracks == nil {
+			continue
+		}
+		if b.Tracks.String() == branch.BranchID.String() {
+			children[branchNode{b.Label, b.BranchID.String()}] = make(tree)
+		}
+	}
+	return children
+}
+
+func (bt *BranchTree) AddBranchFormatting(formatting string) {
+	bt.branchFormatting = formatting
+}
+
+func (bt *BranchTree) AddLocalBranch(branch string) {
+	bt.localBranch = branch
+}
+
+func (bt *BranchTree) AddLocalBranchFormatting(formatting string) {
+	bt.localBranchFormatting = formatting
+}
+
+func (bt *BranchTree) String() string {
+	w := new(bytes.Buffer)
+	var levelsCompleted []int
+	bt.print(w, bt.tree, 0, levelsCompleted)
+	return w.String()
+}
+
+func (bt *BranchTree) print(w io.Writer, currentTree tree, currentLevel int, levelsCompleted []int) {
+	// Sort keys at current level to ensure consistent output
+	var nodes []branchNode
+	for k := range currentTree {
+		nodes = append(nodes, k)
+	}
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].Label < nodes[j].Label
+	})
+
+	for i, node := range nodes {
+		edge := edgeMid
+		if i == len(nodes)-1 {
+			levelsCompleted = append(levelsCompleted, currentLevel)
+			edge = edgeEnd
+		}
+		if bt.isRootNode(node) {
+			edge = ""
+		}
+
+		bt.printNode(w, node, currentLevel, levelsCompleted, edge)
+		if len(currentTree[node]) > 0 {
+			bt.print(w, currentTree[node], currentLevel+1, levelsCompleted)
+		}
+	}
+}
+
+func (bt *BranchTree) printNode(w io.Writer, node branchNode, currentLevel int, levelsCompleted []int, edge string) {
+	// Print necessary edge links for current depth
+	for i := 0; i < currentLevel; i++ {
+		// Apply spacing for completed levels
+		// Do not print edge links for projects with multiple root-level branches
+		if isCompleted(levelsCompleted, i) || (i == 0 && len(bt.rootBranches) > 1) {
```

Comment:
I don't think you need `levelsCompleted` at all. All you need is `currentLevel`, and this conditional would just be `currentLevel == 0`.

Reply:
Sorry this comment wasn't complete. Checking `levelsCompleted` also determines if we should print the link prefix (ie. `|`) for any levels that are still incomplete. Your suggestion would work if it was just spacing as the comment alluded to.